### ISSUE TITLE
twisted.conch.insults.insults.ClientProtocol.dataReceived is really inefficient

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3386]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3386|
|Type|defect|
|Created|2008-08-10 20:58:18Z|
|Branch|https://github.com/twisted/twisted/tree/faster-vt102-client-3386|

This `dataReceived` iterates over a string and, for the application-data case, calls an application method once with each byte of the string.  Performance would be improved non-trivially if runs of application data bytes were batched up and delivered with a single call into application code.

This is similar to the change made to the telnet implementation for [#2107](https://github.com/twisted/twisted/issues/2107).

<details><summary>Searchable metadata</summary>

```
trac-id__3386 3386
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_faster_vt102_client_3386 branches/faster-vt102-client-3386
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1218401898000000 1218401898000000
changetime__1218888226000000 1218888226000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [24571]) Branching to 'faster-vt102-client-3386'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [24572]) bunch up contiguous application data

refs [#3386](https://github.com/twisted/twisted/issues/3386)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

Done, build results: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/faster-vt102-client-3386

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

That's cool. One minor thing: if you're lazy like me and didn't read the wsgi pep, you can get this exception when you don't call start_response:
```
File "/home/therve/projects/Twisted/trunk/twisted/web/wsgi.py", line 256, in _sendResponseHeaders
	    code, message = self.status.split(None, 1)
	exceptions.AttributeError: '_WSGIResponse' object has no attribute 'status'
```

Maybe a meaningfull error like `RuntimeError` would be cool. 

Please merge when you want.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

Blah sorry.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

Nice catch in assertCall. Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [24582]) Merge faster-vt102-client-3386

Author: exarkun
Reviewer: therve
Fixes: [#3386](https://github.com/twisted/twisted/issues/3386)

Change `twisted.conch.insults.insults.ClientProtocol.dataReceived` so that it
buffers up contiguous application data received in a single call and delivers
it to application code all at once instead of delivering it one byte at a time.
This cuts parsing overhead significantly and makes it possible for applications
to handle data more efficiently.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
